### DISCUSSION TITLE
BLE timeout error

### DISCUSSION THREAD

**Discussion by TekyTeky** [Novakasa/brickrail#128] [General]
I just updated to the newest master to test the color sensor build.
I am getting a BLE error in the terminal, then eventually it reverts to the curser and gives a Red error in the GUI.
It was in the Alpha 4 but to a lesser degree, now I cannot Connect Devices. I have tried several ways.

I can connect each device manually, eg Scan for hub, Connect.
But when I hit Connect Devices it fails.
If I try to connect hubs when they are blinking, I get the same issue.

I have attached the logs and text from the terminal.
There are several logs from today.

[brickrail_BLE_Error_logs.zip](https://github.com/Novakasa/brickrail/files/11427341/brickrail_BLE_Error_logs.zip)

[BrickRail BLE error.txt](https://github.com/Novakasa/brickrail/files/11427343/BrickRail.BLE.error.txt)

**Comment by Novakasa** [comment]
It seems like the newly added add_valid_colors func is not found on the train program. Can you redownload the hub programs (activate the Download checkbox in the hub GUI before connecting your hubs) for all the trains you are using and check if the problem still happens?

If that's the case, it is weird that Brickrail didn't detect automatically that the program changed... Did you manually disable the Download checkbox at some point since you updated Brickrail?

**Reply by Novakasa** [reply]
Or are you maybe using multiple hubs with the same name? Then Brickrail might think It already downloaded the new program even though it's a different hub with the same name.

**Reply by Novakasa** [reply]
Also can you send me your .brl layout file? In the logs a (probably unrelated) error shows up that I want to investigate.

**Reply by TekyTeky** [reply]
When I updated to the new master it didn't set the download checkbox.
It had removed it automatically once it does the download.
I had removed 2 trains so when I re added one the Download was set again.

Once I did that and forced the train I didn't remove to download it all worked.

Here is the layout file... it may only hace 2 trains.
[Layout - BAP-Test-BLE.zip](https://github.com/Novakasa/brickrail/files/11429659/Layout.-.BAP-Test-BLE.zip)

**Comment by Novakasa** [comment]
I might have found the reason why it wouldn't trigger the automatic redownload for your hubs. Current master has a fix for it. Next time you run brickrail with the new master version, I would be interested in the brickrail logs.

**Reply by TekyTeky** [reply]
I will soon.

Brickrail logs from this run, selecting download and also the sensor color issue.
[Logs from tonight.zip](https://github.com/Novakasa/brickrail/files/11429722/Logs.from.tonight.zip)

**Reply by TekyTeky** [reply]
This time it set the download flag to update

**Comment by TekyTeky** [comment]
There is an issue now with finding the colors, I get errors in some of them.

In both cases it went over a red sensor.
NOTE- this was the same train/sensor that had the most issues before.. could it be the hub?

![Screenshot 2023-05-09 194236](https://user-images.githubusercontent.com/87248517/237060818-26d72deb-e609-40eb-823f-ee6fab9a318c.png)
![Screenshot 2023-05-09 193030](https://user-images.githubusercontent.com/87248517/237061121-3cbf4d5d-b7d7-4e3b-be88-7a05a83d65c4.png)

**Reply by TekyTeky** [reply]
I changed the Red sensor, that caused the issue to a blue one, like the error was expecting and it worked.

So this red one I changed to red
![Screenshot 2023-05-09 201819](https://github.com/Novakasa/brickrail/assets/87248517/370bc2dd-8746-4ebf-8dd3-16790c529114)

**Reply by Novakasa** [reply]
Did this happen immediately or did it work for a while? If it was working, was it working on Red markers?

**Reply by Novakasa** [reply]
Is this with green baseplates below?

**Reply by TekyTeky** [reply]
I failed immediately with the new master and redownload and once I changed the sensor color then it worked, same for another one that was Red it wanted green, top left. So I did, but then it must have seen a green base plate and then didn't get to blue (top left.)

**Reply by Novakasa** [reply]
Does your layout have any green or yellow markers?

**Reply by TekyTeky** [reply]
No

**Reply by Novakasa** [reply]
BTW, The error message means specifically that according to what you set in Brickrail, the marker it expects is red. However, it measured a different color (green or yellow), so it produces the error.

Currently I am not sure if this is a error in brickrail or if the color measurement is simply unreliable in your layout. If you send me new logs (also a .brl layout file please), I can take a look at it in 5 hours or so.

**Reply by TekyTeky** [reply]
Ok, sure. I will be in bed by then, so take your time.
Also, I tried again and got an assertion error, going from A to B, as it went past the blue one I changed .. C
![Screenshot 2023-05-09 204703](https://github.com/Novakasa/brickrail/assets/87248517/89b6ee18-7417-4c75-b3ee-bd8db1c5c076)

I will get the logs and post then and the layout file below.

**Reply by TekyTeky** [reply]
When I said no to a green sensor, when it asked for a green in top left  I added one. but didn't update the virtual. 

I changed it to Blue and added it to the virtual sensor too and it now works... so blue seems like the best color to get it to work for this train.

**Reply by TekyTeky** [reply]
I replaced all the Red sensors for Blue and the City Train now works.
The Maersk worked on all before.
I re added the red train and it also works.

**Comment by TekyTeky** [comment]
[Recent logs.zip](https://github.com/Novakasa/brickrail/files/11430383/Recent.logs.zip)
[Layout - BAP-Test-BLE-1.zip](https://github.com/Novakasa/brickrail/files/11430386/Layout.-.BAP-Test-BLE-1.zip)

**Reply by Novakasa** [reply]
Could you also upload a picture of the layout?

**Reply by Novakasa** [reply]
And the .brl file?

**Reply by TekyTeky** [reply]
Layout.brl is in the zip above.

**Reply by TekyTeky** [reply]
The layout.brl is in the zip file

**Comment by TekyTeky** [comment]
This first is the B the right blue is the C I mentioned 
![IMG_2488](https://github.com/Novakasa/brickrail/assets/87248517/eb7db469-41af-4a88-81d1-7118142ebea8)

This is the A
![IMG_2489](https://github.com/Novakasa/brickrail/assets/87248517/360c7966-aa1a-4d35-afec-274760caa7b1)

As you can see, I have changed them all to blue now.

**Reply by Novakasa** [reply]
![image](https://github.com/Novakasa/brickrail/assets/59559612/365dc713-0f92-4a42-9b44-0dcf54463871)
Is that also entered as a marker in Brickrail? This will definitely be seen as a red marker, usually, but if you don't have any red markers entered in the virtual layout, this should be ignored.

**Reply by TekyTeky** [reply]
It is a Prior for a siding, which I wasn't testing with... see below.
![SidingRedSensors](https://github.com/Novakasa/brickrail/assets/87248517/198c44ad-7384-4f37-884d-7f010c2d8883)

**Comment by Novakasa** [comment]
I've tweaked the color detection a bit in current master. Also did some light testing using your layout .brl file and couldn't find any issues.

I am still not entirely sure why your color detection was off, but maybe these changes improved it.

I also noticed that reflective materials are quite bad for the sensor. For that reason I recommend regular plates instead of tiles for the markers. Does your layout have any tiles on the tracks? Maybe the switches indeed are the problem, I only use 3D printed ones from trixbrix. The official ones seem to have quite reflective parts. If you still get marker errors, please try to increase the "chroma threshold" setting for your trains to 4000 or so.

Did you use similar switches for the test layout last week?

If this doesn't improve your situation, I might need to do some more extensive color testing and calibration.

**Reply by Novakasa** [reply]
I also added some more print statements for marker detection, so now I would be interested in the BLEServer terminal log as well, as it might now contain some info about what might be wrong in your case.

**Reply by Novakasa** [reply]
In the logs directory there should now be the ble-server log as well.

**Reply by TekyTeky** [reply]
> For that reason I recommend regular plates instead of tiles for the markers. Does your layout have any tiles on the tracks? Maybe the switches indeed are the problem, I only use 3D printed ones from trixbrix. The official ones seem to have quite reflective parts. If you still get marker errors, please try to increase the "chroma threshold" setting for your trains to 4000 or so.

I did have tiles, but thought the same, so changed them to plates. 
Most of the areas I tested with only had one 2 Switches, one going from A to B and the other for the siding on the far left middle of the pic above, and they all went over it without issue, so I don't think it's them.

It does seem to be the red plates.. but was ok before you did the color changes.

> Did you use similar switches for the test layout last week?

Yes, some are even the same ones. I only have Lego Switches, some are modified. 

>  If you still get marker errors, please try to increase the "chroma threshold" setting for your trains to 4000 or so.

It is really the one main train so I can adjust the chroma.

Can you remove green from the equation... as many people have green baseplates, of varying shades.

I'll update at about lunchtime today. it's 6:30am now.

**Reply by TekyTeky** [reply]
OK, so I couldn't wait, I replaced the entry sensors from blue to red again, these were causing issues on A and B and ran the City train, which had the issues.
It went to A from another block before it, without issue and stopped nicely at the Blue. Then from A to B without issue.
So that seems to have worked.
The trains did the downloaded automatically... but not the switches.

**Reply by TekyTeky** [reply]
I also replaced some red plates with red tiles, and they worked ok too.

**Reply by Novakasa** [reply]
> Can you remove green from the equation... as many people have green baseplates, of varying shades.

The way it currently is _supposed_ to work is that as long as you don't set any green markers in Brickrail, the train should never react to any green colors.

I would have hoped I fixed any problem with the commit a few hours ago. Does the train still report measuring a "green" marker with the newest version?

**Reply by TekyTeky** [reply]
> I would have hoped I fixed any problem with the commit a few hours ago. Does the train still report measuring a "green" marker with the newest version?

All seems ok now.. it doesn't see Green anymore, or Yellow.
I'll do some more testing at lunchtime with reversing etc.
Hopefully it should be ok.

**Reply by TekyTeky** [reply]
I did more at lunchtime and did some Random too and it was all good.
Rarely a train gets stuck in reverse still, but if you disconnect it and turn off the hub and reconnect it is fine, but you need to turn off the hub too, the shifty mouse move doesn't resolve it.
It's ok now I know what to do, but maybe you could add a FAQ to the wiki on these weird things.

**Reply by TekyTeky** [reply]
BTW, the updated red error messages are much better :-)

**Reply by Novakasa** [reply]
> Rarely a train gets stuck in reverse still, but if you disconnect it and turn off the hub and reconnect it is fine, but you need to turn off the hub too, the shifty mouse move doesn't resolve it.

Is that after a marker error?

Next time I have the chance, I definitely plan to do a lot of testing with reversing trains.

**Reply by TekyTeky** [reply]
I think it's more if you stop the sequence and a train has been in reverse, it seems to get stuck. There are no errors and can be any train. Often the train on the virtual layout is part way through a block. With the dot at one end or in the middle and the arrow opposite. I try and change it but it just goes backwards, with no error. It will be good for you to have a play. 

Also I only just realised that you can only drag to another block, in Edit mode, when the train changes colour to yellow. My trains are set to light green and sometimes I didn't see the difference and didn't click the train correctly and ended up drawing more track. So a higher contrast between them would be good, plus add how to select the train in the wiki? 

- [ ] Validate train positions
Before operating the layout, you will need to make sure that each train position is consistent with the virtual layout. You can right click + drag each train in edit mode to a new logical block to teleport it to a new position.

Maybe add here to click so it changes colour?

**Reply by Novakasa** [reply]
I think there is currently a bug where the mouse input is not handled correctly when clicking on a selected train.

**Reply by TekyTeky** [reply]
I did more tests, and if you clearly left click on a train so it changes shade, it drags ok, but often you click the block by mistake and it will draw from or near the block.
Since clicking a block in edit mode has no function why not just make the train selection bigger, eg pixel width.. sorry I don't know enough about mouse clicks.. but you know what I mean.

I'll keep an eye out on that bug you created and test, as it kinda drives me crazy. 🤓

**Reply by Novakasa** [reply]
On master I just fixed the issue #131 where when a train was selected, you would simply click "through" the train, selecting what was behind it. I'm not sure if that already addresses all of your reports, we can tweak the input further if required.

**Reply by TekyTeky** [reply]
Ok, thanks. I have today off, so plan to do a lot on the layout. 
I’m let you know for sure.

**Reply by TekyTeky** [reply]
I loaded the Master, should the download be auto selected, as it wasn't?

**Reply by Novakasa** [reply]
The download should only be auto selected if the new release changed the hub programs, or a hub is added with a name that brickrail didn't see before.

**Reply by Novakasa** [reply]
There was another issue related to mouse clicks, the area where you could click a train always had the train wagons with no rotation. So if the train was oriented vertically, it would register the clicks in a weird way (especially the area between wagons). I fixed this in master earlier.

**Comment by TekyTeky** [comment]
Well ignore what I wrote below as now, after sitting for a few hrs and a windows update, it seems to be ok. But maybe it comes and goes.

There is something weird about the Connect Hubs function.
Disconnect Hubs and the closing with X work fine.
But if you start the hubs and then select Connect Hubs it does odd things trying to connect the hubs. They connect and disconnect then wait and you can manually connect them but only some will. Then it sometimes ends the BLE connection.

It's by no way important as I've learnt to connect them manually one by one.
But when you are ready, let me know how you want me to test and what logs etc. or better still try it yourself when you go home, unless you can test with what you have there.

**Comment by TekyTeky** [comment]
Hi, 
I started up with 2 trains & 1 hub connecting manually.
Selected Control Devices All... it was loading then crashed with what seems a BLE error.
It has done this every now and then.
Second restart was fine.
Attached is a zip of all the logs and layout etc.
[Crash_Logs.zip](https://github.com/Novakasa/brickrail/files/11485384/Crash_Logs.zip)

**Comment by TekyTeky** [comment]
Twice in sucession the BLE has disconnected while it was connecting the train and hub.. I've attached the logs.
The third time it worked.
[BLE_Disconnect .zip](https://github.com/Novakasa/brickrail/files/11570751/BLE_Disconnect.zip)

**Reply by Novakasa** [reply]
I'm not sure whether the ble-server.log is from the same run, since there seem to be no issues there, but I do see some errors in the brickrail.log that I fixed on master yesterday. Not sure if these actually cause the problem, since they should probably have caused the issue every time, but to be sure you can just get the newest master build.

**Reply by Novakasa** [reply]
I do see some connection issues myself from time to time, so I will have to investigate more what can be done about them.

**Reply by TekyTeky** [reply]
Yeah it was the correct log, as I grabbed it immeditley, but even if it was from the previos session, that also failed.
The other thing is that the download ticks are there even after downloading and saving the layout, and unticking.
This wasn't like that in the previous master. I downloded this oneon May at 7:24am Sydney time. 
The version says 1.0.0-alpha.4-devc1e55d01, but I guess that isn't the latest master.
I'll download the latest now and recheck things.

**Reply by TekyTeky** [reply]
This new master doesn't seem to keep the Tick active.
